### PR TITLE
[HttpKernel] Create responses for unhandled HttpExceptionInterface exceptions

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #25844
| License       | MIT
| Doc PR        | -

Alternative to #26138

### CODE REVIEW

File: src/Symfony/Component/HttpKernel/HttpKernel.php [hunk 1]
```diff
@@ -230,14 +230,16 @@ private function handleException(\Exception $e, $request, $type)
         // a listener might have replaced the exception
         $e = $event->getException();
 
-        if (!$event->hasResponse()) {
+        if ($event->hasResponse()) {
+            $response = $event->getResponse();
+        } elseif ($e instanceof HttpExceptionInterface) {
+            $response = new Response($e->getMessage());
```
Comment (dmaicher) [inline src/Symfony/Component/HttpKernel/HttpKernel.php]: Wondering... couldn't `$e->getMessage()` potentially expose something that is not meant to be displayed to a user ever?

Maybe better use `Response::$statusTexts[$code]`?

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/HttpKernel.php]: When twig is installed, this is the message that is displayed, so I don't think there is anything harmful here.

Comment (dmaicher) [reply src/Symfony/Component/HttpKernel/HttpKernel.php]: Hmm really? I thought with twig we only show the `status_text`?

https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/TwigBundle/Controller/ExceptionController.php#L66

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/HttpKernel.php]: correct, fixed

**Review by dmaicher — APPROVED** [review]

**Comment by cilefen** [comment]
@nicolas-grekas This is more elegant than #26138.

**Comment by nicolas-grekas** [comment]
Rebased on master, ready for 4.1

**Comment by fabpot** [comment]
Can you add a note in the CHANGELOG?

**Comment by nicolas-grekas** [comment]
CHANGELOG updated

File: src/Symfony/Component/HttpKernel/HttpKernel.php [hunk 2]
```diff
@@ -221,14 +221,17 @@ private function handleException(\Exception $e, Request $request, int $type): Re
         // a listener might have replaced the exception
         $e = $event->getException();
 
-        if (!$event->hasResponse()) {
+        if ($event->hasResponse()) {
+            $response = $event->getResponse();
+        } elseif ($e instanceof HttpExceptionInterface) {
+            $code = $e->getStatusCode();
+            $response = new Response(isset(Response::$statusTexts[$code]) ? $code.' '.Response::$statusTexts[$code] : $code);
```
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/HttpKernel.php]: Would it make sense to do this only in non-debug mode? And keep the exception in debug mode?

**Comment by nicolas-grekas** [comment]
> Would it make sense to do this only in non-debug mode? And keep the exception in debug mode?

HttpKernel doesn't know about debug/non-debug.
But actually, you made me realize this is not the correct approach.
The front controller needs to be updated instead.
I'm going to submit a PR on the recipe.

**Comment by nicolas-grekas** [comment]
Well, adding a listener is possible, nothing prevents it.
But using ExceptionHandler effectively fixes the linked issue as it is described.
We could reopen #26138 for sure, but then it will I'd make it simpler: it should just be a low priority exception listener that, when the event has no reponse set,  uses `ExceptionHandler` as much as possible.
We should not design something that makes it harder for user to handle exceptions themselves (like having to unregister this fallback listener, as the linked PR does for now, and shouldn't have to.)
WDYT?

**Comment by dmaicher** [comment]
> We should not design something that makes it harder for user to handle exceptions themselves (like having to unregister this fallback listener, as the linked PR does for now, and shouldn't have to.)
WDYT?

I agree that this "fallback listener" should only set a response if there is none set yet (by any other listener with higher priority).

But if TwigBundle is registered then it makes sense to remove this fallback listener? or not? Otherwise its just a no-op listener that is instantiated and called for no reason?

**Comment by nicolas-grekas** [comment]
We could remove it, but it should not be a requirement. And in fact, it would be a micro-optimization, since listeners are lazy and exceptions are rare.

**Comment by nicolas-grekas** [comment]
Let's continue on #26138

File: src/Symfony/Component/HttpKernel/HttpKernel.php (hunk 2 above)
Comment (robfrawley) [reply src/Symfony/Component/HttpKernel/HttpKernel.php]: :+1: for keeping them in debug

Comment (robfrawley) [reply src/Symfony/Component/HttpKernel/HttpKernel.php]: Nevermind my prior comment. You already answered it in https://github.com/symfony/symfony/pull/26928#issuecomment-383074122. Please ignore. :-)

**Pull request closed without merging**